builtin/logical/transit/policy.go

```diff
@@ -21,6 +22,173 @@ const (
 	kdfMode = "hmac-sha256-counter"
 )
 
+// policyCache implements a simple locking cache of policies
+type policyCache struct {
+	cache map[string]*lockingPolicy
+	lock  sync.RWMutex
+}
+
+// getPolicy loads a policy into the cache or returns one already in the cache
+func (p *policyCache) getPolicy(req *logical.Request, name string) (*lockingPolicy, error) {
+	// We don't defer this since we may need to give it up and get a write lock
+	p.lock.RLock()
+
+	// First, see if we're in the cache -- if so, return that
+	if p.cache[name] != nil {
+		defer p.lock.RUnlock()
+		return p.cache[name], nil
+	}
+
+	// If we find anything, we'll need to write into the cache, plus possibly
+	// persist the entry, so lock the cache
+	p.lock.RUnlock()
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	// Check one more time to ensure that another process did not write during
+	// our lock switcheroo.
+	if p.cache[name] != nil {
+		return p.cache[name], nil
+	}
+
+	// Note that we don't need to create the locking entry until the end,
+	// because the policy wasn't in the cache so we don't know about it, and we
+	// hold the cache lock so nothing else can be writing it in right now
+
+	// Check if the policy already exists
+	raw, err := req.Storage.Get("policy/" + name)
+	if err != nil {
+		return nil, err
+	}
+	if raw == nil {
+		return nil, nil
+	}
+
+	// Decode the policy
+	policy := &Policy{
+		Keys: KeyEntryMap{},
+	}
+	err = json.Unmarshal(raw.Value, policy)
+	if err != nil {
+		return nil, err
+	}
+
+	persistNeeded := false
+	// Ensure we've moved from Key -> Keys
+	if policy.Key != nil && len(policy.Key) > 0 {
+		policy.migrateKeyToKeysMap()
+		persistNeeded = true
+	}
+
+	// With archiving, past assumptions about the length of the keys map are no longer valid
+	if policy.LatestVersion == 0 && len(policy.Keys) != 0 {
+		policy.LatestVersion = len(policy.Keys)
+		persistNeeded = true
+	}
+
+	// We disallow setting the version to 0, since they start at 1 since moving
+	// to rotate-able keys, so update if it's set to 0
+	if policy.MinDecryptionVersion == 0 {
+		policy.MinDecryptionVersion = 1
+		persistNeeded = true
+	}
+
+	// On first load after an upgrade, copy keys to the archive
+	if policy.ArchiveVersion == 0 {
+		persistNeeded = true
+	}
+
+	if persistNeeded {
+		err = policy.Persist(req.Storage)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	lp := &lockingPolicy{
+		policy: policy,
+	}
+	p.cache[name] = lp
+
+	return lp, nil
+}
+
+// generatePolicy is used to create a new named policy with a randomly
+// generated key
+func (p *policyCache) generatePolicy(storage logical.Storage, name string, derived bool) (*lockingPolicy, error) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	// Ensure one doesn't already exist
+	if lp := p.cache[name]; lp != nil {
```

Comment:
Good point -- this wasn't a problem when we were eager loading but it is now. I'll fix this.